### DISCUSSION TITLE
Can't get crontab to work

### DISCUSSION THREAD

**Discussion by mlinsell** [ZuinigeRijder/SolisCloud2PVOutput#14] [Q&A] [answered]
Hi, 

I have followed the Raspberry pi Configuration steps but have my "solis" folder in home/pi/solis i.e. the users home folder. 

I can run the python script manually ok, but cannot get it scheduled via crontab. I have tried changing the path references to include pi in crontab & solis.sh but still can't get it to work and can't find any error logs. I am new to all this so can someone point me in the direction of what I am doing wrong?

Thanks.

**Comment by ZuinigeRijder** [comment]
Can you post the output of the command: 

````
crontab -l
````
And also the contents of solis.sh

And can you give a directory listing output of the following command:
````
ls -l  ~/solis
````

**Reply by mlinsell** [reply]
Here is the output of crontab -l

> 2 5 * * * ~/pi/solis/solis.sh > /dev/null
> @reboot sleep 123 && ~/pi/solis/solis.sh > /dev/null

And contents of solish.sh 

> script_name=$(basename -- "$0")
> cd ~/pi/solis
> 
> now=$(date)
> if pidof -x "$script_name" -o $$ >/dev/null;then
>    echo "$now: $script_name already running" >> solis.crontab.log 2>&1
>    exit 1
> fi
> echo "WARNING: $now: $script_name NOT running" >> solis.crontab.log 2>&1
> /usr/bin/python -u ~/pi/solis/soliscloud_to_pvoutput.py >> solis.log 2>&1

 

And here is output of ls -l  ~/solis

> 2022.5.3SolisCloudMonitoringAPI1.1.1.pdf
> CHANGELOG.md
> CODE_OF_CONDUCT.md
> LICENSE
> README.md
> 'SolisCloud Monitoring API1.2.pdf'
> soliscloud_to_pvoutput.cfg
> soliscloud_to_pvoutput.py
> solis.sh

**Comment by ZuinigeRijder** [comment]
@mlinsell 

Can you redirect the error logging to a file, e.g. like this in the crontab:
````
2 5 * * * ~/solis/solis.sh >> ~/solis/solis.errors.log 2>&1
@reboot sleep 123 && ~/solis/solis.sh >> ~/solis/solis.errors.log 2>&1
````

Then you maybe see the problem in the solis.errors.log file.

And can you also show the full ls -l output, because then I can see if the permissions are correct.

**Reply by mlinsell** [reply]
Here is the full ls -l output:

> total 1812
-rw-r--r-- 1 pi pi  751784 Mar 27 08:47  2022.5.3SolisCloudMonitoringAPI1.1.1.pdf
-rw-r--r-- 1 pi pi    3263 Mar 27 08:47  CHANGELOG.md
-rw-r--r-- 1 pi pi    5226 Mar 27 08:47  CODE_OF_CONDUCT.md
-rw-r--r-- 1 pi pi   11357 Mar 27 08:47  LICENSE
-rw-r--r-- 1 pi pi    4612 Mar 27 08:47  README.md
-rw-r--r-- 1 pi pi 1046447 Mar 27 08:47 'SolisCloud Monitoring API1.2.pdf'
-rw-r--r-- 1 pi pi     252 Mar 27 09:03  soliscloud_to_pvoutput.cfg
-rw-r--r-- 1 pi pi    8755 Mar 27 09:03  soliscloud_to_pvoutput.py
-rw-r--r-- 1 pi pi     705 Mar 29 17:05  solis.sh

**Comment by ZuinigeRijder** [comment]
Ah, the problem is that the script solis.sh is not executable. 
````
chmod +x ~/solis/solis.sh 
````

These are mine permissions:
````
-rwxr-xr-x 1 rick rick   840 Aug 19  2022 solis.sh
````

**Comment by mlinsell** [comment]
Thanks. I must have missed that when I tried reinstalling it. I have now made the script solis.sh executable and have manually run it from the command and it's creating a log with the following error:

> File "/home/pi/solis/soliscloud_to_pvoutput.py", line 49
    def execute_request(url, data, headers) -> str:
                                            ^
SyntaxError: invalid syntax

**Comment by ZuinigeRijder** [comment, marked as answer]
@mlinsell which version of Python are you running?

````
python --version
````

You need at least python 3.9 or higher. Is it also possible you have python and python3 installed?
You stated that on the commandline it is working, which command did you give then? With python3 in the commandline?
If that is the case, you should adapt the solis.sh script by changing python into python3, e.g.

````
/usr/bin/python3 -u ~/pi/solis/soliscloud_to_pvoutput.py >> solis.log 2>&1
````

**Reply by mlinsell** [reply]
Thanks @ZuinigeRijder 

I upgraded to python 3.11 and this fixed the issue. I had multiple versions of python installed as default with raspberry pi so in case anyone else has issues I would recommend installing the latest version and then adapting the solis.sh script to call the latest version of python you just installed directly like this:

> /usr/bin/python3.11 -u ~/solis/soliscloud_to_pvoutput.py >> ~/solis/solis.log 2>&1

This avoids issues with other apps such as Thonny not working if you replace python3 completely.
